src/listing.rs

```diff
@@ -0,0 +1,252 @@
+            let _ = write!(
+                body,
+                "<tr><td><a class=\"root\" href=\"{}\">..</a></td><td></td></tr>",
+                parent.display()
+            );
+        }
+    }
+
+    let mut entries: Vec<Entry> = Vec::new();
+
+    for entry in dir.path.read_dir()? {
+        if dir.is_visible(&entry) {
+            let entry = entry.unwrap();
+            let p = match entry.path().strip_prefix(&dir.path) {
+                Ok(p) => base.join(p),
+                Err(_) => continue,
+            };
+            // show file url as relative to static path
+            let file_url =
+                utf8_percent_encode(&p.to_string_lossy(), DEFAULT_ENCODE_SET).to_string();
+            // " -- &quot;  & -- &amp;  ' -- &#x27;  < -- &lt;  > -- &gt;
+            let file_name = escape_html_entity(&entry.file_name().to_string_lossy());
+
+            // if file is a directory, add '/' to the end of the name
+            if let Ok(metadata) = entry.metadata() {
+                if skip_symlinks && metadata.file_type().is_symlink() {
+                    continue;
+                }
+                if metadata.is_dir() {
+                    entries.push(Entry::new(file_name, EntryType::Directory, file_url, None));
+                } else {
+                    entries.push(Entry::new(
+                        file_name,
+                        EntryType::File,
+                        file_url,
+                        Some(ByteSize::b(metadata.len())),
+                    ));
+                }
+            } else {
+                continue;
+            }
+        }
+    }
+
+    match sort_method {
+        SortingMethods::Natural => entries
+            .sort_by(|e1, e2| alphanumeric_sort::compare_str(e1.name.clone(), e2.name.clone())),
+        SortingMethods::Alpha => {
+            entries.sort_by(|e1, e2| e1.entry_type.partial_cmp(&e2.entry_type).unwrap());
+            entries.sort_by_key(|e| e.name.clone())
+        }
+        SortingMethods::DirsFirst => {
+            entries.sort_by_key(|e| e.name.clone());
+            entries.sort_by(|e1, e2| e1.entry_type.partial_cmp(&e2.entry_type).unwrap());
+        }
+    };
+
+    if reverse_sort {
+        entries.reverse();
+    }
+
+    for entry in entries {
+        match entry.entry_type {
+            EntryType::Directory => {
+                let _ = write!(
+                    body,
+                    "<tr><td><a class=\"directory\" href=\"{}\">{}/</a></td><td></td></tr>",
+                    entry.link, entry.name
+                );
+            }
+            EntryType::File => {
+                let _ = write!(
+                    body,
+                    "<tr><td><a class=\"file\" href=\"{}\">{}</a></td><td>{}</td></tr>",
+                    entry.link,
+                    entry.name,
+                    entry.size.unwrap()
+                );
+            }
+        }
+    }
+
+    let html = format!(
+        "<html>\
```

Comment:
Side note: What do you think about using `typed-html` here instead of the giant raw string? Not that we _need_ JSX but the typed stuff would be nice.

Reply:
That's a cool idea, would be easier to maintain. Should I include this on this PR ?

Reply:
Nah, let's make it a new one. Keep this one nice and tight.